src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  let newState = {};
+  const historyState = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...copyState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        newState = { ...copyState };
+        action.keysToRemove.forEach((key) => delete newState[key]);
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        newState = copyState;
+        break;
```

Comment:
Using a `switch` statement here is appropriate, but you should not assign `newState` inside each case. Instead, modify `newState` directly and push a copy of it to `historyState` at the end of the loop to avoid repetition.